pkg/build/builder/docker.go

```diff
@@ -171,6 +185,78 @@ func (d *DockerBuilder) addBuildParameters(dir string) error {
 	return nil
 }
 
+// invalidCmd rerpesents an error returned from replaceValidCmd
+// when an invalid Dockerfile command has been passed to
+// replaceValidCmd
+var invalidCmd = errors.New("invalid Dockerfile command")
+
+// replaceValidCmd replaces the valid occurrence of cmd
+// in a Dockerfile with the given replaceStr
+func replaceValidCmd(cmd, replaceStr string, fileData []byte) (string, error) {
+	if _, ok := dockercmd.Commands[cmd]; !ok {
+		return string(fileData), invalidCmd
+	}
+	buf := bytes.NewBuffer(fileData)
+	// Parse with Docker parser
+	node, err := parser.Parse(buf)
+	if err != nil {
+		return string(fileData), errors.New("cannot parse Dockerfile")
+	}
+
+	var pos int
+	switch cmd {
+	case dockercmd.From, dockercmd.Entrypoint, dockercmd.Cmd:
+		pos = traverseAST(cmd, node)
+	default:
+		return string(fileData), nil
+	}
+
+	// Re-initialize the buffer
+	buf = bytes.NewBuffer(fileData)
+	var newFileData string
+	var index int
+	for {
+		line, err := buf.ReadString('\n')
+		if err != nil {
+			if err == io.EOF {
+				break
+			}
+			return string(fileData), err
+		}
+		if imageRegex.MatchString(line) {
+			index++
+			if index == pos {
+				line = imageRegex.ReplaceAllLiteralString(line, replaceStr)
+			}
+		}
+		newFileData += line
+	}
+	return newFileData, nil
+}
+
+// traverseAST traverses the Abstract Syntax Tree output
+// from the Docker parser and returns the valid position
+// of the command it was requested to look for
+func traverseAST(cmd string, node *parser.Node) int {
```

Comment:
Yep, you are right. I wanted to add a sanity check inside that function but since it's recursive it would probably affect performance a bit. So I agree with you about documenting it and I will address it.